librad/src/git/local/transport.rs

```diff
@@ -30,55 +30,54 @@ use crate::{
     git::{
         ext::{into_git_err, RECEIVE_PACK_HEADER, UPLOAD_PACK_HEADER},
         local::{self, url::LocalUrl},
-        storage::{self, Storage},
+        storage::{self, Storage, WithSigner},
     },
-    keys::SecretKey,
+    keys,
     paths::Paths,
     uri::RadUrn,
 };
 
-lazy_static! {
-    static ref SETTINGS: Arc<RwLock<Option<Settings>>> = Arc::new(RwLock::new(None));
```

Comment:
@kim Not sure if this was a safe change, but it seemed to work without everything exploding. Also tried calling `register` twice in the `propagation_test` and it seemed to be fine.

I remember you saying it was necessary for the transport registration, but can't remember why this setup exactly.

See what you think when you get back.

Reply:
Ya we need the global variable to be able to reconfigure at runtime, eg with different paths (for testing).

Reply:
My thoughts are that a global variable would be the same as exposing the Arc + RwLock (changes here https://github.com/radicle-dev/radicle-link/pull/255/commits/07afd1ff6033e8e0b54e1112b778905ccf11a61c). The only thing is that it's exposed to the caller. But I'm not sure how else to get around this :/

Reply:
It’s not the same — registering the transport with libgit2 happens only once, and is stored in a static in C. So when the transport gets invoked, it’s on the struct that is stored at that location — creating and modifying a new LocalTransport has no effect.

Reply:
Well bollix... I tried all the `Box<dyn Signer>` magic I, and others, could think of without making the compiler :weary: I'm not sure how to get a statically know `Signer` without choosing a concrete type. The only concrete type being the `SecretKey`.

Reply:
How so? A Box<dyn > works fine here: https://github.com/radicle-dev/radicle-link/blob/master/librad/src/git/p2p/transport.rs#L68

Reply:
I can write up a commit to show you tomorrow.

But it boils down to needing `Clone` for `Storage` (because of the `TryCow` stuff). And then our `dyn Signer<Error = Box<dyn StdError>` isn't cloneable.

Reply:
Ah I see. rustc clearly wins the award for most misleading type errors.

Storage doesn’t care which signer its type parameter is instantiated to, only if there is one (hence the name WithSigner, could also be SomeSigner or AnySigner). If I’m not mistaken, you should be able to instantiate S to either Box<dyn Signer + Clone>, or a newtype around that. A Signer should always be Clone, so perhaps we can add that as a bound on the trait itself.

Reply:
Ah heh, right the second pain in the ass is that you can't do Box<dyn Signer + Clone> because it gives out about not being allowed to have more than one non-auto trait.

Reply:
Seems like we're pretty fucked here. So I started fleshing out the `Box` approach again. To recap we have the following blanket trait:
```rust
pub trait Signer: sign::Signer + Send + Sync + Clone + 'static {}
```

So I went back to `local/transport.rs` and changed `LocalTransport` to:
```rust
type BoxedSigner = Box<dyn Signer<Error = Box<dyn std::error::Error + Send + Sync + 'static>>>;

#[derive(Clone)]
pub struct LocalTransport {
    storage: Arc<Mutex<Storage<BoxedSigner>>>,
}
```

To which I get the following error:
```
error[E0038]: the trait `git::storage::Signer` cannot be made into an object
   --> librad/src/git/local/transport.rs:236:5
    |
236 |     storage: Arc<Mutex<Storage<BoxedSigner>>>,
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `git::storage::Signer` cannot be made into an object
    | 
   ::: librad/src/git/storage.rs:154:48
    |
154 | pub trait Signer: sign::Signer + Send + Sync + Clone + 'static {}
    |           ------                               ----- ...because it requires `Self: Sized`
    |           |
    |           this trait cannot be made into an object...

error: aborting due to previous error

For more information about this error, try `rustc --explain E0038`.
```

I look into the error code from `rustc --explain E0038` and this is the first snippet:
```
Trait objects like `Box<Trait>` can only be constructed when certain
requirements are satisfied by the trait in question.

Trait objects are a form of dynamic dispatch and use a dynamically sized type
for the inner type. So, for a given trait `Trait`, when `Trait` is treated as a
type, as in `Box<Trait>`, the inner type is 'unsized'. In such cases the boxed
pointer is a 'fat pointer' that contains an extra pointer to a table of methods
(among other things) for dynamic dispatch. This design mandates some
restrictions on the types of traits that are allowed to be used in trait
objects, which are collectively termed as 'object safety' rules.

Attempting to create a trait object for a non object-safe trait will trigger
this error.

There are various rules:

### The trait cannot require `Self: Sized`

When `Trait` is treated as a type, the type does not implement the special
`Sized` trait, because the type does not have a known size at compile time and
can only be accessed behind a pointer. Thus, if we have a trait like the
following:

```
trait Foo where Self: Sized {

}
```

We cannot create an object of type `Box<Foo>` or `&Foo` since in this case
`Self` would not be `Sized`.

Generally, `Self: Sized` is used to indicate that the trait should not be used
as a trait object. If the trait comes from your own crate, consider removing
this restriction.
```

Right, I don't have a sized on `Signer`, but guess who does?! [`Clone`](https://doc.rust-lang.org/src/core/clone.rs.html#110-134) does.

Reply:
Ach ach, das kompilern von die kostet-nix abstrakten ist nicht vor die gefingerpoken und mittengrabben! Gefahr von das hidden cost!

I’m not sure why you need to tell it at all what the associated error type is, but we can change `keystore::Signer` as we like, no? Just slap `Sized + Send + Sync` there (it will have to have those bounds anyway), and refine to `Clone` here. Does that not work?

Reply:
> I’m not sure why you need to tell it at all what the associated error type is

Der Compiler fragt danach

> but we can change keystore::Signer as we like, no? Just slap Sized + Send + Sync there (it will have to have those bounds anyway)

Do you mean in `radicle-keystore`? I dunno how that helps? 

The above explanation, as far as I'm understanding, says that any trait with `Self: Sized` can't be packed as a `Box<dyn Trait>`. We want a `Box` because of this global variable. We want a `Clone` because of the `Store` And `Clone` wants a `Self: Sized`. If we pass these conditions into the Galois SMT solver I think it won't be solveable :upside_down_face:

Reply:
Sometimes it seems they’re just making shit up, and then forget why they ended up there.

I believe it might work to use `ToOwned` instead of `Clone` for `S`.

Reply:
`ToOwned` wasn't going to work either because then we would need to say what `Owned` would be.

The latest [commit](https://github.com/radicle-dev/radicle-link/pull/255/commits/a081b64d7cfca03626344bb98e1bcd56335f5e1a) works with a double-heelflip-body-varial-to-revert, but it depends on this change https://github.com/radicle-dev/radicle-keystore/pull/5.

Please let this saga be over...

Reply:
Turns out DynClone was enough.

NOW, let this saga be over.